Index: typo3/sysext/cms/tslib/class.tslib_fe.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- typo3/sysext/cms/tslib/class.tslib_fe.php	(revision c66ae06f0a2ceaebae09c2a184076f954e9dd4ef)
+++ typo3/sysext/cms/tslib/class.tslib_fe.php	(revision )
@@ -811,7 +811,7 @@
 			}
 
 				// The preview flag will be set if a backend user is in an offline workspace
-			if (($GLOBALS['BE_USER']->user['workspace_preview'] || t3lib_div::_GP('ADMCMD_view') || $this->doWorkspacePreview()) && ($this->whichWorkspace()===-1 || $this->whichWorkspace()>0))	{
+			if (($GLOBALS['BE_USER']->user['workspace_preview'] || $this->doWorkspacePreview()) && ($this->whichWorkspace()===-1 || $this->whichWorkspace()>0))	{
 				$this->fePreview = 2;	// Will show special preview message.
 			}
 
@@ -920,7 +920,7 @@
 
 			// Initialize the page-select functions.
 		$this->sys_page = t3lib_div::makeInstance('t3lib_pageSelect');
-		$this->sys_page->versioningPreview = ($this->fePreview===2 || intval($this->workspacePreview) || t3lib_div::_GP('ADMCMD_view')) ? TRUE : FALSE;
+		$this->sys_page->versioningPreview = ($this->fePreview===2 || intval($this->workspacePreview) ) ? TRUE : FALSE;
 		$this->sys_page->versioningWorkspaceId = $this->whichWorkspace();
 		$this->sys_page->init($this->showHiddenPage);
 
