Index: typo3conf/ext/templavoila/mod1/index.php
===================================================================
--- typo3conf/ext/templavoila/mod1/index.php	(revision 197)
+++ typo3conf/ext/templavoila/mod1/index.php	(working copy)
@@ -2529,8 +2529,14 @@
 							$params = '&edit[pages]['.intval($this->id).']=edit';
 						}
 						if ($params) {
-							$returnUrl = '&returnUrl='.rawurlencode(t3lib_extMgm::extRelPath('templavoila').'mod1/index.php?'.$this->link_getParameters());
-							$redirectLocation = $GLOBALS['BACK_PATH'].'alt_doc.php?'.$params.$returnUrl;	//.'&localizationMode=text';
+               //patch for aio_feedit
+               if ( t3lib_div::_GP("returnUrlFeedit") ){
+                  $returnUrl = '&returnUrl='.t3lib_div::_GP("returnUrlFeedit");
+               }else{
+                 $returnUrl = '&returnUrl='.rawurlencode(t3lib_extMgm::extRelPath('templavoila').'mod1/index.php?'.$this->link_getParameters());
+               }
+               //patch for aio_feedit end
+               $redirectLocation = $GLOBALS['BACK_PATH'].'alt_doc.php?'.$params.$returnUrl;  //.'&localizationMode=text';
 						}
 					break;
 				}
